historycombobox.cpp

```diff
@@ -0,0 +1,74 @@
+/**
+ * \file
+ * <!--
+ * Copyright 2015 Develer S.r.l. (http://www.develer.com/)
+ * -->
+ *
+ * \brief HistoryComboBox class implementation
+ *
+ * \author Aurelien Rainone <aurelien@develer.org>
+ */
+
+#include "historycombobox.h"
+#include <history.h>
+#include <algorithm>
+#include <QLineEdit>
+
+
+HistoryComboBox::HistoryComboBox(QWidget *parent) :
+    QComboBox(parent)
+{
+    history = new History(this);
+
+    connect(this, static_cast<void (HistoryComboBox::*)(int)>(&HistoryComboBox::activated),
+            history, &History::setCurrent);
+}
+
+void HistoryComboBox::fillList(QString current_text)
+{
+    QStringList hist_lines(history->getHistory());
+
+    clear();
+    addItems(hist_lines);
+    setCurrentIndex(hist_lines.indexOf(current_text));
+}
+
+void HistoryComboBox::keyPressEvent(QKeyEvent *e)
+{
+    switch (e->key())
+    {
+        case Qt::Key_Up:
+            // history current not set means we just added an element or we
+            // added nothing yet. either case there is no more recent element
+            if (history->current() != -1)
+            {
+                setCurrentIndex(0);
+                setItemText(0, history->previous());
+                fillList(this->lineEdit()->text());
+            }
+            break;
+
+        case Qt::Key_Down:
+            setCurrentIndex(0);
+            setItemText(0, history->next());
+            fillList(this->lineEdit()->text());
+            break;
+
```

Comment:
Credo che questi due casi qua sopra (`Qt::Key_Up`e `Qt::Key_Down`) debbano essere invertiti: nella shell, se non stai navigando la history e premi la freccia "up", ti appare il comando più recente, mentre col codice che hai scritto non succede niente (perché siamo in fondo alla coda della history).

Può essere che quando hai invertito l'ordinamento della lista della history tu dovessi passare anche di qui per completare la modifica?